### DISCUSSION TITLE
How does SimulatedClock work in tests?

### DISCUSSION THREAD

**Discussion by douglaseggleton** [statelyai/xstate#2545] [Q&A]
I have a simple state machine for an 💡 **Energy Efficient LightBulb**.
The behaviour is a on/off `TOGGLE_SWITCH` and an automatic turn-off after 5 seconds. 

```ts
export default createMachine({
  initial: "off",
  states: {
    on: {
      on: {
        TOGGLE_SWITCH: {
          target: "off"
        },
      },
      after: {
        5000: {
          target: "off"
        }
      }
    },
    off: {
      on: {
        TOGGLE_SWITCH: {
          target: "on"
        }
      }
    }
  }
});
```

I am trying to create a set of tests to test the behaviour of the machine - In particular, the timing event has the following test:

### ✅   Passes

```ts
test("it should turn off after 5 seconds", (done) => {
  const service = interpret(lightBulbMachine, {
    clock: new SimulatedClock()
  }).onTransition((state, event) => {
    if (state.matches("off") && event.type !== "TOGGLE_SWITCH") {
      done();
    }
  });
  const clock = service.clock as SimulatedClock;

  service.start();
  service.send("TOGGLE_SWITCH");
  clock.increment(6000);
});
```

My understanding is to start the service and send events until the machine reaches the desired state.
I specifically want to check that the lightbulb was turned off as a result of the timer and not the toggle (`event.type !== "TOGGLE_SWITCH"`).

## Question

I thought this test could be improved by putting the test in the `on` state by default. However, with the following, the `after.5000` event does not fire and the machine does not transition to `off`.

### ❌  Fails

```ts
  // ...
  service.start('on');
  clock.increment(6000);
```

Is the above behaviour intended or do I misunderstand how to use SimulatedClock in testing?

## Example
https://codesandbox.io/s/lightbulb-fsm-xstate-mpu61?file=/src/lightBulbMachine.test.ts

## Reference Docs
https://xstate.js.org/docs/guides/delays.html#interpretation
https://xstate.js.org/docs/guides/delays.html#interpretation

**Comment by davidkpiano** [comment]
The state passed into `service.start(...)` needs to be the complete state to get the desired effect, _including actions_.

The simplest way to do this, given your machine, is to "get" to the proper `'on'` state:

```js
const onState = lightBulbMachine.transition(undefined, 'TOGGLE_SWITCH'); // transition from initial to 'on' state

// ...

service.start(onState);
clock.increment(6000);
```
